File: 98.cpp
#include <stack>
#include <vector>
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
using namespace std;
class Solution {
public:
  bool isValidBST(TreeNode *root) {
    stack<TreeNode *> s;
    vector<int> ans;
    TreeNode *node = root;
    while (node != nullptr || !s.empty()) {
      while (node != nullptr) {
        s.push(node);
        node = node->left;
      }
      node = s.top();
      s.pop();
      ans.push_back(node->val);
      node = node->right;
    }
    for (int i = 0; i < ans.size() - 1; i++) {
      if (ans[i+1] <= ans[i]) {
        return false;
      }
    }
    return true;
  }
};

File: 98.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    arr = []
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.inOrder(root)
        s = sorted(self.arr)
        return all(self.arr[i] == s[i] for i in range(len(s)))
    def inOrder(self, root: Optional[TreeNode]) -> None:
        if root:
            self.inOrder(root.left)
            self.arr.append(root.val)
            self.inOrder(root.right)


if __name__ == "__main__":
    t = TreeNode(0)
    s = Solution()
    print(s.isValidBST(t))
